File: app/src/main/java/me/lukeforit/spaceofaday/ui/base/DIViewModelFactory.java
package me.lukeforit.spaceofaday.ui.base;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import java.util.Map;

import javax.inject.Inject;
import javax.inject.Provider;

public class DIViewModelFactory implements ViewModelProvider.Factory {

    private final Map<Class<? extends ViewModel>, Provider<ViewModel>> viewModelProviderMap;

    @Inject
    public DIViewModelFactory(@NonNull Map<Class<? extends ViewModel>, Provider<ViewModel>> viewModelProviderMap) {
        this.viewModelProviderMap = viewModelProviderMap;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) viewModelProviderMap.get(modelClass).get();
    }
}

File: app/src/main/java/me/lukeforit/spaceofaday/ui/base/DIViewModelFactory.kt
package me.lukeforit.spaceofaday.ui.base

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import javax.inject.Inject
import javax.inject.Provider

class DIViewModelFactory @Inject
constructor(private val viewModelProviderMap: Map<Class<out ViewModel>, @JvmSuppressWildcards Provider<ViewModel>>) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return viewModelProviderMap[modelClass]!!.get() as T
    }
}
